lightning/src/ln/channel.rs

```diff
@@ -1305,8 +1323,12 @@ impl<Signer: Sign> Channel<Signer> {
 			self.holding_cell_htlc_updates.push(HTLCUpdateAwaitingACK::ClaimHTLC {
 				payment_preimage: payment_preimage_arg, htlc_id: htlc_id_arg,
 			});
+			#[cfg(any(test, feature = "fuzztarget"))]
+			self.historical_inbound_htlc_fulfills.insert(htlc_id_arg);
 			return Ok((None, Some(monitor_update)));
 		}
+		#[cfg(any(test, feature = "fuzztarget"))]
+		self.historical_inbound_htlc_fulfills.insert(htlc_id_arg);
```

Comment:
Also add a `contains()` check inside the `HTLCUpdateAwaitingACK::ClaimHTLC` branch in `get_update_fail_htlc` to pass the debug_assert in case of downgrade from `update_fulfill_htlc`to `update_fail_htlc`, which is also a correct behavior ?

Reply:
We always `debug_assert!(false, ...` there, because its a bug to duplicate-fail (as failing waits for the commitment signed dance to complete).

Reply:
I was thinking about the sequence success-failure? We don't wait on forward channel full commitment_signed_dance to pass success backward, so the `get_update_fail_htlc` in `process_pending_htlc_forwards` can be reached second time ?

Reply:
Oh! Yep, you're totally right, worse the fuzzer won't hit that case. I updated the new test in this PR to hit that case and fixed bugs.